### PR TITLE
Guided Tours: Documentation on future architecture work

### PR DESCRIPTION
**Pull request by mcsf**
I wanted to gather some thoughts and put them somewhere accessible, as reducing any [bus factor](https://en.wikipedia.org/wiki/Bus_factor) is always good.

### CODE REVIEW

File: client/layout/guided-tours/docs/ARCHITECTURE-FUTURE.md [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+Every time `actionLog` changes, Guided Tours's main selectors get called with the new state, whether a tour is running or not. This has obvious **performance implications**. Right now, we believe this is acceptably mitigated by 1) copious use of memoization with `createSelector` and simple cache heuristics, and 2) the still limited set of action types `actionLog` subscribes to.
+
+Nevertheless, whenever the selectors have to be fully reevaluated, all of `actionLog` has to be considered anew: all those sub-selectors will perform multiple `map`s, `reduce`s, etc. on the latest `actionLog`, even though most of its items have already been mapped, reduced, etc. in previous versions of `actionLog`.
+
+Now, you know what they say about premature optimization. Which is why we haven't optimized yet. But this document seeks to gather thoughts for the day when, through meticulous profiling (best-case scenario) or out of desperate hunches (worst-), we decide optimization is needed. Next comes a collection of ideas on how to address the issue, some of them composable, some mutually incompatible.
```
Comment (lsinger) [inline client/layout/guided-tours/docs/ARCHITECTURE-FUTURE.md]: Don't assume people are familiar with that saying.

File: client/layout/guided-tours/docs/ARCHITECTURE-FUTURE.md [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+### Streams!
+
+You probably saw this one coming. Calypso does not use [streams][rxjs], but, should we decide in favor of them, an improved `actionLog` would be had at minimum cost: as mentioned in _§ On the nature of `actionLog`_, the RxJS operators would naturally be adequate and optimized to handle `actionLog`. Realistically, though, adopting RxJS is much more than just pulling a large library (compressed/uncompressed: 30.1/138 KB), it's adopting a whole new programming paradigm. As much as I'd like to see that happen, it probably won't anytime soon.
+
+### (Persistent?) Data Structures
+
+This approach is a poor man's substitute for streams. It aims to replace standard array traversal functions with a suite of functions optimized for `actionLog`: `mapActionLog`, `filterActionLog`, etc. Since `actionLog` is a [monoid][monoids], any operation _f_ on `actionLog` can be decomposable to [divide and conquer]. Notably,
```
Comment (lsinger) [inline client/layout/guided-tours/docs/ARCHITECTURE-FUTURE.md]: Could you rephrase? "Using persistent data structures could be a simple replacement for streams ..."

File: client/layout/guided-tours/docs/ARCHITECTURE-FUTURE.md [hunk 3]
```diff
@@ -0,0 +1,150 @@
+# Guided Tours: Architecture considerations for the future
+
```
Comment (lsinger) [inline client/layout/guided-tours/docs/ARCHITECTURE-FUTURE.md]: Add an explainer sentence here what this document is about and who would need or want to read it in what situation. Similar to your PR description should be fine.

**Comment by lsinger** [comment]
Left a few comments. Once those are addressed I think this is GTG.

**Comment by mcsf** [comment]
@lsinger thanks for the good feedback! Updated.
